Migration to Mojo 0.25.6

moseschmiedel opened the pull request:
See https://docs.modular.com/mojo/changelog#v0256-2025-09-22

moseschmiedel committed:
Overload `unsafe_set` to work with pointer to component data array

When doing batch operations it is more efficient to `memcpy` the whole component data array from the old archetype to the new archetype, instead of copying every component instance one by one.

moseschmiedel committed:
Test if archetypes have correct entity amount

moseschmiedel committed:
Debug printing

moseschmiedel committed:
Add StaticVariant

StaticVariant can model a value that can have multiple types but it is known at compile-time which type will be used.
It may be helpful to think of it as the Variant type from stdlib, but resolved completely at compile-time.

moseschmiedel committed:
Implement common traits for StaticVariant when wrapped type implements them

moseschmiedel committed:
Add batch component removal

moseschmiedel committed:
Add test for batch component removal

moseschmiedel committed:
Change batch addition test to process more entities

moseschmiedel committed:
Change batch addition test to process more entities

moseschmiedel committed:
Align docstring for batch `world.remove` to docstring for batch `world.add`

moseschmiedel committed:
Fix assertions in `test_world_batch_remove` to use entity methods

moseschmiedel committed:
Move entities during `world.remove` and update `EntityIndex`

moseschmiedel committed:
Extend batch addition benchmark to cover batch removal

moseschmiedel committed:
Add 1000x1000 batch add & remove benchmark

moseschmiedel committed:
Fix import statement and use new DType.uint

moseschmiedel committed:
Remove `run_destructors` parameter from `InlineArray`

moseschmiedel committed:
Introduce `ImplicitlyCopyable` trait and remove `hint_trivial_type` from `List`

moseschmiedel committed:
Use 'var' instead of 'owned', `sizeof` renamed to `size_of`

moseschmiedel committed:
Add `deinit` keyword to `__moveinit__` constructors

moseschmiedel committed:
Use explicit `.copy()` more often

moseschmiedel committed:
Update indexing

moseschmiedel committed:
Fix indexing in EntityAccessor and cleanup copy for Node

moseschmiedel committed:
`unsafe_take` is not needed anymore, use `VariadicPack.consume_elements` instead

moseschmiedel committed:
`ArchetypeByListIterator` should take an owned List as argument

moseschmiedel committed:
Update Mojo to 0.25.6

moseschmiedel committed:
Don't expose internal detail (archetypes) in user-facing error message

moseschmiedel committed:
Remove duplicate lines in `world_test.mojo`

samufi commented:
I would argue that everything that holds a somewhat constant state should be implicitly copyable and everything else not. So I would see PitPool and _ArchetypeIterator on the explicit side. But we can see if this makes ergonomics bad...

Component is not a struct I believe...

moseschmiedel committed:
Update benchmark code to Mojo 0.25.6

moseschmiedel committed:
Update AoS benchmark Mojo version to 0.25.6

moseschmiedel committed:
`copy` should not be overridden anymore

moseschmiedel committed:
Update examples to Mojo 0.25.6

moseschmiedel committed:
Remove `ImplicitlyCopyable` from `Archetype`

moseschmiedel committed:
Make `Entity` and `EntityIndex` `ImplicitlyCopyable` and make `EntityIndex` `Movable` too

moseschmiedel committed:
Make `LockMask` `ImplicitlyCopyable`

moseschmiedel commented:
> I would argue that everything that holds a somewhat constant state should be implicitly copyable and everything else not. So I would see PitPool and _ArchetypeIterator on the explicit side. But we can see if this makes ergonomics bad...

If we allow `ImplicitCopyable` on `BitPool` we can let the compiler synthesize the `__copyinit__` for `LockMask` for us. Otherwise we need to write it ourselves. Does that change your opinion on that?

moseschmiedel committed:
Make `Query` and `QueryInfo` `ImplicitlyCopyable`

moseschmiedel committed:
Remove `ImplicitlyCopyable` from `StaticOptional`, make `StaticVariant` `Copyable & Movable`

samufi commented:
> > Component is not a struct I believe...
> 
> You are right, it is an alias to `Copyable & Movable` at the moment. The question is, if we require `ImplicitlyCopyable` here aswell?

I would require as little as possible here, unless it enforces correctness somehow. So if there was a trait for trivial types I would use that. But ImplicitlyCopyable does not suffice, so I would leave it away.

samufi commented:
> > I would argue that everything that holds a somewhat constant state should be implicitly copyable and everything else not. So I would see PitPool and _ArchetypeIterator on the explicit side. But we can see if this makes ergonomics bad...
> 
> If we allow `ImplicitCopyable` on `BitPool` we can let the compiler synthesize the `__copyinit__` for `LockMask` for us. Otherwise we need to write it ourselves. Does that change your opinion on that?

I actually think it is the other way round. LockMask should not be implicitly copyable, as this is the one instance that contains all the locks. The name somehow suggests that it is simply some kind of boolean array, which is why I thought it should be trivital. Maybe refactoring here could be good. Maybe to `Locker` or `Lock`?

moseschmiedel committed:
Remove `ImplicitlyCopyable` from `ComponentType`

moseschmiedel committed:
Rename `LockMask` to `LockManager` and remove `ImplicitlyCopyable`

moseschmiedel committed:
Remove redundant `Raises:` from doc-string

moseschmiedel committed:
Update changelog

moseschmiedel committed:
v0.4.0 is already released

moseschmiedel committed:
Add explicit origin parameter for ref in `test_utils.{load,store}`

Implicit origin of ref is triggering a crash of `mojo doc` (see https://github.com/modular/modular/issues/5361).

moseschmiedel committed:
Upgrade Modo version

moseschmiedel committed:
Update changelog

moseschmiedel committed:
`Dict.__getitem__` returns mutable ref now

samufi commented on changelog.md:
```diff
@@ -3,11 +3,42 @@
 
 ### Breaking changes
+- Update the utilized Mojo version to 25.6 and adjust the code accordingly.
+- Revisit what structs are only `Copyable` and which can be also `ImplicitlyCopyable`
```
Revisit *which* ...

samufi commented on changelog.md:
```diff
@@ -3,11 +3,42 @@
+  - Entity
+  - EntityIndex
+  - LockedContext
+  - LockMask -> LockManager
```
This is not implicitly copyable, is it?

moseschmiedel replied:
No, it is not.

samufi commented on changelog.md:
```diff
@@ -3,11 +3,42 @@
 ### Other changes
 - Implement batch component addition as overload of `world.add`
 - Implement batch component removal as overload of `world.remove`
+- Remove `unsafe_take`
```
from the _utils module

samufi commented on src/larecs/bitmask.mojo:
```diff
@@ -191,12 +193,12 @@ struct BitMask(Copyable, EqualityComparable, KeyElement, Movable, Stringable):
     fn contains_any(self, other: Self) -> Bool:
         """Reports if any bit of the other mask is in this mask."""
-        return ((self._bytes & other._bytes) != 0).reduce_or()
+        return (self._bytes & other._bytes) != 0
```
Isn't this doing an implicit reduce_and?

moseschmiedel replied:
No, first `self._bytes & other._bytes` results in a new `SIMD[DType.uint8, Self.total_bytes]` where only the bits are 1 that occur in both `BitMask`s. Comparing this vector to 0 is equal to asking if any bit is 1 which in turn is only the case if the bit was 1 in `self._bytes` and `other._bytes`.

samufi replied:
OK, understood. == becomes reduce_and and != becomes reduce_or.

samufi commented on src/larecs/query.mojo:
```diff
@@ -102,13 +113,11 @@ struct Query[
         """
         iterator = self._world[]._get_entity_iterator(
-            self._mask, self._without_mask
+            self._mask, self._without_mask.copy()
```
Maybe we should make _get_entity_iterator take the without_mask not as owned. I think there is actually no real benefit of it being copied. (Also, it makes the code ugly)

moseschmiedel replied:
I think it is good to leave this decision to the user of `_get_entity_iterator`. In the end we always need an owned `without_mask` as `_ArchetypeByMaskIterator` wants one. With the current API the user of `_get_entity_iterator` can decide if he wants to copy some value they already have into the iterator or if they even want to move the value. When we decide to not take this argument as owned, we always have to do the copy internally which could be optimized away in some cases previously.

samufi replied:
To protocol our off-line discussion: in this particular case, we gain ergonomics by making this not owned but do not loose control, as the type is trivial and needs to be copied completely even in move operation.

samufi commented on src/larecs/query.mojo:
```diff
@@ -324,7 +343,7 @@ struct _ArchetypeByMaskIterator[
         query_info = Self.QueryInfo(
             mask=self._mask,
-            without_mask=self._without_mask,
+            without_mask=self._without_mask.copy(),
```
Maybe also here let us not use an owned mask but rather copy the provided value in the constructor.

moseschmiedel replied:
Same argument as with `_get_entity_iterator`.

moseschmiedel replied:
Removed the owned requirement here as well to increase ergonomics.

samufi commented on src/larecs/resource.mojo:
```diff
@@ -57,57 +48,57 @@ struct Resources(ExplicitlyCopyable, Movable, Sized):
+            if id in self._storage:
+                conflicting_ids.append(id)
+
+        if len(conflicting_ids) > 0:
```
Could also be done with an implicit bool cast.

samufi commented on src/larecs/resource.mojo:
```diff
@@ -57,57 +48,57 @@ struct Resources(ExplicitlyCopyable, Movable, Sized):
-        if id in self._storage:
-            raise Error("Resource already exists.")
-        self._storage[id] = UnsafeBox(resource[])
+        self._storage[id] = UnsafeBox(resource[].copy())
```
This always makes a copy when the resource is added. Do we find a way to move resources into Reseources (at least if the ^ operator is used)?

samufi replied:
Actually, I never see this overload used. Let us just delete it then.

samufi commented on src/larecs/resource.mojo:
```diff
@@ -122,40 +113,49 @@ struct Resources(ExplicitlyCopyable, Movable, Sized):
 
     @always_inline
     fn _set[
         T: Copyable & Movable, add_if_not_found: Bool
```
Is it possible now to make T infer only, so that we do not have to explicitly write add_if_not_found=add_if_not_found later? (There used to be a bug, but maybe it has been fixed?)

samufi commented on src/larecs/world.mojo:
```diff
@@ -466,8 +427,8 @@ struct World[*component_types: ComponentType](
-                    T = Ts[i.value], assert_has_component=False
-                ](index_in_archetype) = components[i]
+                    T = Ts[i], assert_has_component=False
+                ](index_in_archetype) = components[i].copy()
```
Maybe a dumn question, but assigning a copied value does not copy twice, does it? (like, evelaute the rhs, then copy the value to the lhs)

moseschmiedel replied:
I assume this to behave the same as if I would construct a new Component on the rhs (`archetype[]... = Component(42)`). In this case there is no separate copy, the newly constructed Component is directly initialized in the memory pointed to by the ref on the lhs. Extrapolating from this the explicit copy should also initialize itself directly in the memory pointed to by the ref. See https://docs.modular.com/mojo/manual/lifecycle/life#copy-constructor.

samufi commented on src/larecs/world.mojo:
```diff
@@ -1579,7 +1538,7 @@ struct World[*component_types: ComponentType](
                 Pointer(to=self._archetypes),
                 query.mask,
-                query.without_mask,
+                query.without_mask.copy(),
```
Maybe we can also here take the without_mask not as owned...

samufi commented on src/larecs/world.mojo:
```diff
@@ -1697,9 +1656,9 @@ struct World[*component_types: ComponentType](
-        owned mask: BitMask,
-        owned without_mask: StaticOptional[BitMask, has_without_mask],
-        owned start_indices: _EntityIterator[
+        var mask: BitMask,
```
Let us rethink which of the elements really need to be takes as owned. Otherwise we are spending a long time writing .copy() - at least for the elements that are not implicitly copyable...

moseschmiedel replied:
Going from the previous discussion about ownership of `without_mask` I would also change `mask` to be unowned. `start_indices` should stay owned IMO because it is a `List` which is not cheap to copy.

samufi reviewed (changes requested):
Good job! Only minor things we may want to think about...

moseschmiedel committed:
Add suggestions to `changelog.md`

moseschmiedel committed:
Use implicit Bool cast to check for non-empty `List`

moseschmiedel committed:
Remove dead overload for `Resources._add`

moseschmiedel committed:
Copy `without_mask` always implicitly to increase ergonomics

moseschmiedel committed:
Move `start_indices` in `world._get_entity_iterator` as it is not needed anymore afterwards

moseschmiedel committed:
Infer-only bug seems to be fixed

moseschmiedel committed:
Take `mask: BitMask` not as owned

samufi commented on benchmark/plots/src/aos.mojo:
```diff
@@ -210,7 +210,7 @@ fn run_benchmarks(config: BenchConfig) raises -> List[BenchResult]:
             results.append(result)
 
-    return results
+    return results^
```
Instead of moving the result out, we should use the out argument convention and directly fill the output slot.

samufi commented on src/larecs/resource.mojo:
```diff
@@ -57,57 +48,42 @@ struct Resources(ExplicitlyCopyable, Movable, Sized):
-        for i in range(resources.__len__()):
-            self._add(get_type_name[Ts[i]](), unsafe_take(resources[i]))
-        __disable_del resources
+        conflicting_ids = List[StringSlice[StaticConstantOrigin]]()
```
We could initialize the list with a minimal capacity (0?), as it is unlikely that there are conflicting IDs - and if there are some, an extra allocation is not a problem. Alternatively we could also start by first checking if anything is duplicate and only then allocate a list to return the IDs where this is an issue. I don't think this is a real problem, but please consider it.

moseschmiedel replied:
Yes, I think the current version is more or less already doing what you are suggesting. If I forced the minimal capacity to 0 the first "real" allocation (besides allocating space for the pointer and size of the list, which are on the stack) happens when the first conflicting ID is found. Note that the problem here is not looking for duplicates in `*Ts` but to look for already registered resources in `self._storage`. So to do the check before allocating the list would mean that we need to first iterate through `resources` until we find a conflict and then allocate the list and then continue iterating. This is nearly equivalent to allocating the list with capacity=0, iterating through `resources` and appending all conflicts.

samufi reviewed (changes requested):
Very minor changes.

moseschmiedel committed:
Initialize `conflicting_ids` in `Resources.add` with capacity=0

moseschmiedel committed:
Use `out` argument for `plots/src/aos.mojo:run_benchmarks`

samufi commented on benchmark/world_benchmark.mojo:
```diff
@@ -472,6 +472,56 @@ fn prevent_inlining_add_remove_1_comp() raises:
+        world = SmallWorld()
+
+        # create 1_000_000 entities that initially do not have FlexibleComponent[1]
+        _ = world.add_entities(pos, count=1_000_000)
```
I think the benchmakr below covers this already and we do not need two benchmarks for this. I'd suggest removing it.

samufi commented on benchmark/world_benchmark.mojo:
```diff
@@ -537,11 +587,14 @@ fn benchmark_batch_add_1_comp_1_000_000(
         _ = world.add(
             world.query[Position]().without[FlexibleComponent[1]](), comp
         )
+        _ = world.remove[FlexibleComponent[1]](
```
If I am not mistaken, this does as much add_entities as add and remove. I'd suggest doing the add and remove 1000 times and working with 1000 entities only.

samufi commented on benchmark/world_benchmark.mojo:
```diff
@@ -573,6 +626,75 @@ fn benchmark_batch_add_5_comp_1_000_000(
             comp4,
             comp5,
         )
```
As above: I think it is sufficient to keep the 1000 x 1000 benchmark

samufi commented on AGENTS.md:
```diff
@@ -0,0 +1,31 @@
+- Use snake_case for functions/variables, PascalCase for types
+- Use `fn` for static functions, `def` for dynamic functions
+- Prefer `var` for mutable, immutable by default
+- Use `inout` parameters for mutation, not return modified values
```
Use `mut` not `inout`

samufi commented on AGENTS.md:
```diff
@@ -0,0 +1,31 @@
+
+## Code Style
+- Use snake_case for functions/variables, PascalCase for types
+- Use `fn` for static functions, `def` for dynamic functions
```
For now, always use `fn`. We may use `def` in main and raising test functions.

samufi commented on AGENTS.md:
```diff
@@ -0,0 +1,31 @@
+- Leverage SIMD types `SIMD[type, width]` for vectorization
+- Apply traits: `Copyable`, `Movable`, `Stringable` appropriately
+- Use manual memory management with `UnsafePointer` when needed
+- Include docstrings for public APIs
```
Every function should have a docstring including a description of all parameters, raises and returns.

samufi commented on AGENTS.md:
```diff
@@ -0,0 +1,31 @@
+## Error Handling & Safety
+- Follow borrow checker principles for memory safety
+- Prefer stack allocation and RAII patterns
+- Use `debug_warn()` utility for debug messages
```
maybe also use `debug_assert` for critical checks that should not impact performance

moseschmiedel committed:
Improve `AGENTS.md`

moseschmiedel committed:
Remove redundant benchmarks
